PayloadPkg/OsLoader/OsLoader.c

```diff
@@ -484,6 +484,128 @@ StartBooting (
   return EFI_SUCCESS;
 }
 
+/**
+  Start pre-OS boot image
+
+  This function will jump to pre-OS boot image entry point.
+
+  @param[in]  PreBootLoadedImage   Loaded pre-OS boot image information.
+  @param[in]  OsLoadedImage        Loaded OS boot image information.
+
+  @retval  RETURN_SUCCESS       boot image is return after boot
+  @retval  Others               There is error when checking boot image
+**/
+EFI_STATUS
+StartBootingPreOs (
+  IN LOADED_IMAGE            *PreBootLoadedImage,
+  IN LOADED_IMAGE            *OsLoadedImage
+  )
+{
+  EFI_STATUS                 Status;
+  PRE_OS_BOOT_ENTRY          PreOsBootMain;
+  PRE_OS_BOOT_PARAMS         PreOsBootParams;
+
+  DEBUG_CODE_BEGIN();
+  PrintStackHeapInfo ();
+  DEBUG_CODE_END();
+
+  PreOsBootParams.eaxGpr = 0x0;
+  PreOsBootParams.ecxGpr = 0x0;
+  PreOsBootParams.edxGpr = 0x0;
+  PreOsBootParams.ebxGpr = 0x0;
+  PreOsBootParams.espGpr = 0x0;
+  PreOsBootParams.ebpGpr = 0x0;
+
+  Status = RETURN_SUCCESS;
+
+  DEBUG ((DEBUG_INIT, "Jumping into Pre-OS Boot image entry point...\n"));
+  PreOsBootMain = (PRE_OS_BOOT_ENTRY) PreBootLoadedImage->Image.MultiBoot.BootState.EntryPoint;
+  if (PreOsBootMain == 0) {
+    DEBUG ((DEBUG_ERROR, "Pre-OS Boot EntryPoint is not found\n"));
+    return RETURN_INVALID_PARAMETER;
+  }
+  BeforeOSJump ("Calling Pre-OS Boot Entry ...");
+  PreOsBootMain ((VOID*) &PreOsBootParams);
+  Status = RETURN_DEVICE_ERROR;
+
+  if (Status == RETURN_DEVICE_ERROR) {
+    CpuHalt ("Boot image returned");
+  }
+
+  return RETURN_SUCCESS;
+}
+
+/**
+  Boot pre-OS executable from OsBootOption
+
+  Based on given boot option, this function will load a pre-OS boot image,
+  setup boot parameters for OS boot image, and load OS boot image for pre-OS
+  boot image to launch after pre-OS boot image finishes execution.
+
+  @param[in]  InPreBootOption     Pre-OS boot option to boot
+  @param[in]  InOsBootOption      OS boot option to boot
+
+  @retval  RETURN_SUCCESS       Image returns after boot into its entrypoint
+  @retval  Others               There is error to boot from this boot option
+**/
+EFI_STATUS
+BootPreBootImage (
```

Comment:
This code is very similar to BootOsImage().   Can we consider to enable BootOsImage() to support both OS and PreOS image?

Reply:
Yes. I prefer doing similar with trusty boot. Adding a new OS flag with minor code change.